snippets/2D_geometry/furthest_pair.cpp

```diff
@@ -0,0 +1,25 @@
+#define sq(a) ((a)*(a))
+double dist2(point a) { return sq(real(a))+sq(imag(a)); }
+
+// returns diameter (max pairwise distance) of points on a convex hull.
+// needs that all points are distinct and none are collinear
+// takes time linear in the size of the hull
+pair<point,point> hull_diameter(convex_polygon poly) {
+    vector<point> &S = poly.points;
+    if(S.size() < 2) return {point(), point()};
+    int n = S.size(), j = 1;
+    pair<ll, pair<point,point>> res({0, {S[0], S[0]}});
+    for(int i=0;i<j;i++)
+        for (;; j = (j + 1) % n) {
```

Comment:
can this be a binary search instead?

Reply:
hmm, maybe not: consider a case such as

![image](https://user-images.githubusercontent.com/10997917/97199221-8c34d680-178e-11eb-84dc-ce38b8c846f6.png)

where you can't really binary search a function like that

Reply:
it uses the [method of rotating calipers](https://en.wikipedia.org/wiki/Rotating_calipers)

Reply:
ah, my bad I didn't fully understand the code. That looks good to me